File: services/api/app/domain/post_audience.py
"""F42. The four audiences a post can be addressed to, and who they reach.

Spec section F42 names four levels and stops there. What the spec does not say,
and what this module exists to pin down, is that they are a *vocabulary* and
not a ladder: `friends` and `group` reach disjoint sets of people, and neither
contains the other. Written as a ladder -- the obvious implementation, a rank
comparison like the one in `visibility.py` -- a person who shares a group with
the author reads posts addressed to the author's friends, which is a different
promise from the one the word on the button made.

## Why this module takes facts and not identities

`can_read` is handed `is_friend` and `is_group_member` as booleans that someone
else already proved. It cannot look them up and must not: the whole reason F42
is a privacy feature is that "who are your friends" and "who is in this group"
are answers only the server holds, read from `friend_requests` and
`memberships` at the moment of the read. A version of this function that
accepted a list of friend ids would work identically against a list supplied by
the caller, and the caller is the one person whose claim about their own
membership is worth nothing.

That is also why `friends` is not frozen at write time. Unfriending someone has
to take away what they can see; a recipient list computed when the post was
written would keep them in it forever.

Pure functions over plain dicts. No I/O, no ORM, no framework.
"""

from __future__ import annotations

__all__ = [
    "AUDIENCES",
    "DEFAULT_AUDIENCE",
    "AudienceError",
    "can_read",
    "check_writable",
    "needs_context",
]

#: Narrowest first, purely so the tuple reads in a sensible order. Nothing in
#: this module compares two audiences by position, and nothing should: see the
#: module docstring on why these are not a ladder.
AUDIENCES = ("only_me", "friends", "group", "public")

#: What a client that says nothing gets. The narrowest level that is still a
#: post rather than a note to oneself would be `friends`; this picks
#: `only_me` instead, because the cost of guessing too narrow is that somebody
#: re-posts, and the cost of guessing too wide cannot be taken back.
DEFAULT_AUDIENCE = "only_me"

#: ADR-0022 §2.2: who may comment on a person's posts, chosen by that person.
#: `readers` is everyone `can_read` lets in; `friends` narrows that to friends;
#: `nobody` closes the composer for everyone but the author. The CHECK on
#: `people.wall_comment_policy` is the second spelling of this tuple.
COMMENT_POLICIES = ("readers", "friends", "nobody")
DEFAULT_COMMENT_POLICY = "readers"


class AudienceError(Exception):
    def __init__(self, code: str):
        super().__init__(code)
        self.code = code


def needs_context(audience: str) -> bool:
    """Only `group` names a group. Everything else addresses people."""
    return audience == "group"


def check_writable(audience: str, context_id: object | None) -> None:
    """Refuse a post whose audience and target do not agree.

    Mirrored by the `audience_matches_target` CHECK constraint on the table.
    Two spellings of one rule, deliberately: change either and change both.
    The constraint is what makes the rule true of rows written by anything
    that skips this layer; this is what makes the refusal a 422 instead of an
    IntegrityError surfacing as a 500.
    """
    if audience not in AUDIENCES:
        raise AudienceError("UNKNOWN_AUDIENCE")
    if needs_context(audience) and context_id is None:
        raise AudienceError("GROUP_AUDIENCE_NEEDS_CONTEXT")
    if not needs_context(audience) and context_id is not None:
        # A `only_me` post carrying a group id is a row that looks group-scoped
        # to every future query written by someone who did not read this file.
        raise AudienceError("CONTEXT_NOT_ADDRESSABLE")


def is_comment_policy(value: object) -> bool:
    return isinstance(value, str) and value in COMMENT_POLICIES


def can_comment(
    post: dict,
    *,
    policy: str,
    reader_id: str,
    is_friend: bool,
    is_group_member: bool,
) -> bool:
    """Whether one reader may comment under one post (ADR-0022 §2.2).

    Reading comes first: nobody comments on what they cannot see. Then the
    wall owner's policy -- the author's own, never the commenter's -- with the
    author exempt from their own setting. A policy word this module does not
    know closes the composer, for the same reason `can_read` fails closed on
    an audience it does not know.
    """
    if not can_read(
        post,
        reader_id=reader_id,
        is_friend=is_friend,
        is_group_member=is_group_member,
    ):
        return False
    if reader_id == post.get("author_id"):
        return True
    if policy == "readers":
        return True
    if policy == "friends":
        return bool(is_friend)
    return False


def can_delete_comment(comment: dict, post: dict, actor_id: str) -> bool:
    """The comment's author may take it back; the post's author may sweep
    their own wall. Nobody else -- there is no moderator role."""
    return actor_id == comment.get("author_id") or actor_id == post.get("author_id")


def can_read(
    post: dict,
    *,
    reader_id: str,
    is_friend: bool,
    is_group_member: bool,
) -> bool:
    """Whether one reader may see one post. The only place this is decided.

    `is_friend` means "reader and author are friends *now*", and
    `is_group_member` means "reader is in the group this post names *now*".
    Both are the caller's to prove, and both are ignored for audiences they do
    not apply to -- that asymmetry is the rule the `only_me` tests are about.
    """
    audience = post.get("audience")
    if audience not in AUDIENCES:
        # Fail closed, the same way `can_view_history` does. An audience string
        # this module does not recognise is a fact nobody proved, and reading a
        # missing fact as permissive is how a schema change becomes a leak.
        return False

    # The author is the one reader every audience includes, `only_me` included.
    # Checked first so no audience branch has to remember to allow it.
    if reader_id == post.get("author_id"):
        return True

    if audience == "public":
        return True
    if audience == "friends":
        return is_friend
    if audience == "group":
        # `is_group_member` is a claim about the group this post names. With no
        # group named there is nothing for it to be a claim about, so it is not
        # evidence of anything and the post stays with its author.
        return post.get("context_id") is not None and is_group_member
    # `only_me`, and it reached here, so the reader is not the author.
    return False

File: services/core/internal/domain/postaudience/postaudience.go
// Package postaudience ports app.domain.post_audience (F42, ADR-0022 §2.2,
// ADR-0023 §2.3.2): the four audiences a post can address, who they reach, who
// may comment, and what a block hides.
//
// Parity, not correctness, is the contract (ADR-0029 §2.4): oracle_test.go
// replays testdata/python_*.json, rendered by
// scripts/render_domain_w2_goldens.py from the real module in the parity API
// image.
//
// # Values
//
// The post dict is what the service builds (`_post_dict`): the author id, the
// audience word and the context id or None. The facts are booleans somebody
// else proved. is_comment_policy takes any object in Python; the port takes
// the str the service passes.
package postaudience

import "slices"

// The four audiences (AUDIENCES), narrowest first.
const (
	OnlyMe  = "only_me"
	Friends = "friends"
	Group   = "group"
	Public  = "public"
)

// DefaultAudience is DEFAULT_AUDIENCE.
const DefaultAudience = OnlyMe

// The comment policies (COMMENT_POLICIES).
const (
	PolicyReaders = "readers"
	PolicyFriends = "friends"
	PolicyNobody  = "nobody"
)

// DefaultCommentPolicy is DEFAULT_COMMENT_POLICY.
const DefaultCommentPolicy = PolicyReaders

// Codes AudienceError carries.
const (
	CodeUnknownAudience           = "UNKNOWN_AUDIENCE"
	CodeGroupAudienceNeedsContext = "GROUP_AUDIENCE_NEEDS_CONTEXT"
	CodeContextNotAddressable     = "CONTEXT_NOT_ADDRESSABLE"
)

var (
	audiences       = [...]string{OnlyMe, Friends, Group, Public}
	commentPolicies = [...]string{PolicyReaders, PolicyFriends, PolicyNobody}
)

// Audiences returns AUDIENCES in declaration order.
func Audiences() []string { return slices.Clone(audiences[:]) }

// CommentPolicies returns COMMENT_POLICIES in declaration order.
func CommentPolicies() []string { return slices.Clone(commentPolicies[:]) }

// AudienceError is Python's AudienceError: `str(exc)` is the code.
type AudienceError struct {
	Code string
}

func (e *AudienceError) Error() string { return e.Code }

// Post is the post dict. ContextID nil is None.
type Post struct {
	AuthorID  string
	Audience  string
	ContextID *string
}

// Comment is the comment dict can_delete_comment reads.
type Comment struct {
	AuthorID string
}

// NeedsContext is needs_context.
func NeedsContext(audience string) bool {
	return audience == Group
}

// CheckWritable is check_writable. Any non-nil context, the empty string
// included, names a group.
func CheckWritable(audience string, contextID *string) error {
	if !slices.Contains(audiences[:], audience) {
		return &AudienceError{Code: CodeUnknownAudience}
	}
	if NeedsContext(audience) && contextID == nil {
		return &AudienceError{Code: CodeGroupAudienceNeedsContext}
	}
	if !NeedsContext(audience) && contextID != nil {
		return &AudienceError{Code: CodeContextNotAddressable}
	}
	return nil
}

// IsCommentPolicy is is_comment_policy for a str.
func IsCommentPolicy(value string) bool {
	return slices.Contains(commentPolicies[:], value)
}

// CanComment is can_comment. An unknown policy closes the composer for
// everybody but the author.
func CanComment(post Post, policy, readerID string, isFriend, isGroupMember bool) bool {
	if !CanRead(post, readerID, isFriend, isGroupMember) {
		return false
	}
	if readerID == post.AuthorID {
		return true
	}
	switch policy {
	case PolicyReaders:
		return true
	case PolicyFriends:
		return isFriend
	}
	return false
}

// VisibleTo is visible_to: CanRead, then the block rule. A block hides
// everything but a group post, both ways; the author always reads their own.
func VisibleTo(post Post, readerID string, isFriend, isGroupMember, isBlocked bool) bool {
	if !CanRead(post, readerID, isFriend, isGroupMember) {
		return false
	}
	if !isBlocked || readerID == post.AuthorID {
		return true
	}
	return post.Audience == Group
}

// CanDeleteComment is can_delete_comment.
func CanDeleteComment(comment Comment, post Post, actorID string) bool {
	return actorID == comment.AuthorID || actorID == post.AuthorID
}

// CanRead is can_read: an unknown audience fails closed, even for the author.
func CanRead(post Post, readerID string, isFriend, isGroupMember bool) bool {
	if !slices.Contains(audiences[:], post.Audience) {
		return false
	}
	if readerID == post.AuthorID {
		return true
	}
	switch post.Audience {
	case Public:
		return true
	case Friends:
		return isFriend
	case Group:
		return post.ContextID != nil && isGroupMember
	}
	return false
}
